File: jan-15-max-area-of-square-hole/solution.js
/**
 * @param {number} n
 * @param {number} m
 * @param {number[]} hBars
 * @param {number[]} vBars
 * @return {number}
 */
var maximizeSquareHoleArea = function(n, m, hBars, vBars) {
  const maxConsecutive = (arr) => {
    if (!arr || arr.length === 0) return 0;
    arr.sort((a, b) => a - b);
    let best = 1;
    let cur = 1;
    for (let i = 1; i < arr.length; i++) {
      if (arr[i] === arr[i - 1] + 1) {
        cur++;
      } else {
        if (cur > best) best = cur;
        cur = 1;
      }
    }
    if (cur > best) best = cur;
    return best;
  };

  const h = maxConsecutive(hBars) + 1;
  const v = maxConsecutive(vBars) + 1;
  const side = Math.min(h, v);
  return side * side;
};

File: jan-15-max-area-of-square-hole/solution.java
import java.util.Arrays;

class Solution {
    public int maximizeSquareHoleArea(int n, int m, int[] hBars, int[] vBars) {
        int h = maxConsecutive(hBars) + 1;
        int v = maxConsecutive(vBars) + 1;
        int side = Math.min(h, v);
        return side * side;
    }

    private int maxConsecutive(int[] arr) {
        if (arr == null || arr.length == 0) return 0;
        Arrays.sort(arr);
        int best = 1;
        int cur = 1;
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] == arr[i - 1] + 1) {
                cur++;
            } else {
                if (cur > best) best = cur;
                cur = 1;
            }
        }
        if (cur > best) best = cur;
        return best;
    }
}
